File: web/src/api/weather.ts
import { apiClient } from "./client";

type WeatherForecastResponse = {
  daily: {
    time?: string[];
    weather_code?: number[];
    temperature_2m_max?: number[];
    temperature_2m_min?: number[];
  };
};

export type FetchWeatherForecastOptions = {
  signal?: AbortSignal;
  /** When true, server bypasses its successful forecast TTL cache. */
  force?: boolean;
};

export function fetchWeatherForecast(
  location: string,
  startDate: string,
  endDate: string,
  options?: FetchWeatherForecastOptions | AbortSignal,
): Promise<WeatherForecastResponse> {
  const normalized =
    options instanceof AbortSignal ? { signal: options } : (options ?? {});
  return apiClient.get<WeatherForecastResponse>(
    "/api/v1/weather/forecast",
    {
      location,
      startDate,
      endDate,
      ...(normalized.force ? { force: "true" } : {}),
    },
    { signal: normalized.signal },
  );
}

File: server/api/schemas/responses/weather.py
"""Weather forecast response models."""

from __future__ import annotations

from pydantic import BaseModel, Field


class WeatherDailyResponse(BaseModel):
    time: list[str] = Field(default_factory=list)
    weather_code: list[int] = Field(default_factory=list)
    temperature_2m_max: list[float] = Field(default_factory=list)
    temperature_2m_min: list[float] = Field(default_factory=list)


class WeatherForecastResponse(BaseModel):
    daily: WeatherDailyResponse
